### PR TITLE
Hosting Command Palette: Split the search into label and keywords to display command labels first 

### PR DESCRIPTION
**Pull request by katinthehatsite**
Closes https://github.com/Automattic/dotcom-forge/issues/4836

## Proposed Changes

This PR improves the search for the commands:

* if there is a search keyword that matches the label, it will match and display that command to the top:

Screenshot 2023-12-13 at 3 49 05 PM

* if there is a search keyword that matches the `searchLabel`, it will add these commands next:

Screenshot 2023-12-13 at 3 50 35 PM

* the filtering on the site will remain as is as per the issue description

## Testing Instructions

* Navigate to Calypso Live Link
* Press `cmd + k` on macOS or `ctrl + k` on Windows to trigger command palette 
* Type `reader` in the search field and observe that the command that comes up is `Open reader`
* Type `edge` in the search field and observe that the commands that come up first have `edge` in the command label and the commands that come after are the commands that have `edge` in their `searchLabel`
* Navigate to the nested commands
* Confirm that the sites filtering remains unaffected and works as before

**Notes**
* I am wondering if we should use the `filter` prop that is [already with the `cmdk`](https://www.npmjs.com/package/cmdk#command-cmdk-root) and try to explain the logic there or we should stick with our custom logic - what are your thoughts @sejas ?

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/components/command-palette/use-command-palette.tsx [hunk 1]
```diff
@@ -74,73 +75,94 @@ const siteToAction =
...
-			const hasNoContext = commandHasContext( b.context );
-
-			if ( hasContextCommand && ! hasNoContext ) {
-				return -1; // commands with context come first if there is a context match
-			} else if ( ! hasContextCommand && hasNoContext ) {
-				return 1; // commands without context set
+	// Logic for selected command (sites)
+	if ( selectedCommandName ) {
+		const selectedCommand = commands.find( ( c ) => c.name === selectedCommandName );
+		let sitesToPick = null;
```
Comment (sejas) [inline client/components/command-palette/use-command-palette.tsx:94]: Thanks for the refactor!
Now it's more clear that this hook is running two different algorithms for sites and commands.

File: client/components/command-palette/use-command-palette.tsx [hunk 2]
```diff
@@ -74,73 +75,94 @@ const siteToAction =
...
+			command.label.toLowerCase().includes( searchLowerCase )
+		);
+		const commandSearchLabelMatch = commands.filter(
+			( command ) =>
+				command.searchLabel?.toLowerCase().includes( searchLowerCase ) &&
+				! commandLabelMatch.includes( command )
+		);
+
+		// Combine label and searchLabel matches
+		finalSortedCommands = [ ...commandLabelMatch, ...commandSearchLabelMatch ];
```
Comment (sejas) [inline client/components/command-palette/use-command-palette.tsx:134]: I like the approach of filtering first by `label` and then by `searchLabel` 👍

**Review by sejas — COMMENTED** [review]
@katinthehatsite , thanks for the quick PR and the optimizations of `useCommandPalette` hook. I like how now it separates the logic for sites and commands 👍 .

I'll make a deeper review tomorrow. It's working as expected though 👌 . Great job!


https://github.com/Automattic/wp-calypso/assets/779993/77e9d91c-ff3e-4669-92c2-4f8c2a5cf2e7

File: client/components/command-palette/use-command-palette.tsx [hunk 3]
```diff
@@ -74,73 +75,94 @@ const siteToAction =
...
+
+		// Initialize the final result with sorted commands
+		finalSortedCommands = [ ...sortedCommands ];
+
+		// Add the "viewMySites" command to the beginning in all contexts except "/sites"
+		if ( viewMySitesCommand && currentPath !== '/sites' ) {
+			finalSortedCommands.unshift( viewMySitesCommand );
 		}
-		sitesToPick = filteredSites.map( siteToAction( onClick ) );
 	}
```
Comment (danielbachhuber) [inline client/components/command-palette/use-command-palette.tsx:165]: It'd be great to have some tests around all of this sorting/filtering behavior. It's hard to assess whether fixing one scenario is breaking another scenario.

Comment (katinthehatsite) [reply client/components/command-palette/use-command-palette.tsx:165]: 100%, I was thinking about this already - @sejas I don't know how fast you would like to merge. Should we do a separate issue to add the tests for both sorting and filtering or should we handle it as a part of this PR?

Comment (sejas) [reply client/components/command-palette/use-command-palette.tsx:165]: Let's handle the tests as a separate PR.
I'll let the issue open until we have implemented the tests.

**Review by sejas — COMMENTED** [review]
I found an error on `cmdk` that seems difficult to fix.
When changing the order, sometimes the commands scroll moves down.
Investigating a bit, seems that `cmdk` scrolls to the previous active command.
I tried to fix it with `useMemo` and also by scrolling to the top when the search or commands length changes, and It doesn't always work.


https://github.com/Automattic/wp-calypso/assets/779993/faa2dcb8-141c-4cbe-b4ce-bac5682759ea



Steps to reproduce:
1. Search `add`
2. Delete letter by letter
3. When you have deleted the last letter you will observe that the command's scroll has moved.

> I am wondering if we should use the filter prop that is [already with the cmdk](https://www.npmjs.com/package/cmdk#command-cmdk-root) and try to explain the logic there or we should stick with our custom logic - what are your thoughts @sejas ?

I think you are right and we should explore the native `filter` prop.
Inspecting how it works, we may need to concat `label` and `sublabel` with a "special/magic" separator to then split it to rank higher the matches with the label. 

I like the optimization implemented in this PR that returns the sitesToPick without sorting the commands. We can transform this PR into that optimization.

File: client/components/command-palette/use-command-palette.tsx (hunk 3 above)
Comment (katinthehatsite) [reply client/components/command-palette/use-command-palette.tsx:165]: I added an issue here to create sorting and filtering tests: https://github.com/Automattic/dotcom-forge/issues/4952

**Comment by katinthehatsite** [comment]
Thanks for super detailed testing @sejas ! I suggest we proceed with the following steps:

* to keep the optimization for when the nested commands are executed, I created a PR here that we can merge independently of this PR: https://github.com/Automattic/wp-calypso/pull/85386
* I also created the issue for adding tests for sorting here: https://github.com/Automattic/dotcom-forge/issues/4952 
* as for this specific issue, let's explore the built-in `filter` approach. I will leave this PR as open for now and we can explore the `filter` in a separate PR and compare and see if we run into the same issue

How does this sound?

**Comment by sejas** [comment]
@katinthehatsite , thank you for moving the optimization in a separate PR and creating the issue for the tests 🙏 .
Let's give a higher priority to the built-in `filter` over writing the tests.

**Comment by katinthehatsite** [comment]
I will close it in favour of https://github.com/Automattic/wp-calypso/pull/85418 since we decided to go with a different approach

**Pull request closed without merging**